src/Printers/Languages/PrinterCommandSet.ts

```diff
@@ -0,0 +1,109 @@
+    /** Gets the command to end a document. */
+    abstract get documentEndCommand(): Uint8Array;
+    /** Gets the command language this command set implements */
+    abstract get commandLanugage(): PrinterCommandLanguage;
+
+    /** Transpile a document into a document ready to be sent to the printer. */
+    transpileDoc(doc: Commands.IDocument): Readonly<Commands.CompiledDocument> {
+        const validationErrors = [];
+        const state = new Commands.CompiledDocument(this.commandLanugage);
+        state.addRawCmd(this.documentStartCommand);
+        doc.commands.forEach((c) => {
+            try {
+                state.addRawCmd(this.transpileCommand(c, state));
+                state.commandEffectFlags |=
+                    c.printerEffectFlags ?? Commands.PrinterCommandEffectFlags.none;
+            } catch (e) {
+                if (e instanceof DocumentValidationError) {
+                    validationErrors.push(e);
+                } else {
+                    throw e;
+                }
+            }
+        });
+        state.addRawCmd(this.documentEndCommand);
+        if (validationErrors.length > 0) {
+            throw new DocumentValidationError('One or more validation errors', validationErrors);
+        }
+        return Object.freeze(state);
+    }
+
+    /**
+     * Round a raw value to the nearest step.
+     */
+    protected roundToNearestStep(value: number, step: number): number {
+        const inverse = 1.0 / step;
+        return Math.round(value * inverse) / inverse;
+    }
+
+    /** Strip a string of invalid characters for a command. */
+    protected cleanString(str: string) {
+        return str
+            .replace(/\\/gi, '\\\\')
+            .replace(/"/gi, '\\"')
+            .replace(/[\r\n]+/gi, ' ');
+    }
+
+    /** Start a new label within the same command batch. */
+    protected startNewDocument() {
+        return this.combineCommands(this.documentEndCommand, this.documentStartCommand);
+    }
+
+    /** Apply an offset command to a document. */
+    protected modifyOffset(cmd: Commands.Offset, outDoc: Commands.CompiledDocument) {
+        const newHoriz = cmd.absolute ? cmd.horizontal : outDoc.horizontalOffset + cmd.horizontal;
+        outDoc.horizontalOffset = newHoriz < 0 ? 0 : newHoriz;
+        if (cmd.vertical) {
+            const newVert = cmd.absolute ? cmd.vertical : outDoc.verticalOffset + cmd.vertical;
+            outDoc.verticalOffset = newVert < 0 ? 0 : newVert;
+        }
+        return this.noop;
+    }
+
+    /** Combine two commands into one command array. */
+    protected combineCommands(cmd1: Uint8Array, cmd2: Uint8Array) {
+        const merged = new Uint8Array(cmd1.length + cmd2.length);
+        merged.set(cmd1);
+        merged.set(cmd2, cmd1.length);
+        return merged;
+    }
+
+    /** Transpile a command to its native command equivalent. */
+    abstract transpileCommand(cmd: IPrinterCommand, outDoc: CompiledDocument): Uint8Array;
```

Comment:
Basically, because data strings exist.

For raw data, such as images, fonts, or other things, we need the ability to skip past the encode operation. Frustratingly this is often a mixture of text-to-encoded-text bits, such as the command parameters, and then the raw data embedded within it. Some of them include multiple runs of raw data separated by multiple commas for added fun.

So _most_ commands need to be encoded entirely, and _some_ commands are a mixture of stuff to encode and stuff to not encode. Thus individual transpile operations must handle their own encoding rules.

Ultimately transpiling is taking the class structure data and turning it into the command representation of what gets sent to the printer. Each transpile operation controls that, and the encode _utility method_ is at the ready to help with that operation. That's the reason they're separate.

The changes in #22 altered this relationship quite a bit, more of the heavy lifting is done within the PrinterCommandSet class and less in the Epl and Zpl command set classes. It might be more clear where these two methods interplay now.